File: src/j/implementation/microSet/InnerArraySet.java
package j.implementation.microSet;

import j.generator.Generator;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.Set;

/**
 * Created by fly on 2/23/17.
 */
public class InnerArraySet<T> extends ArrayList<T> implements InnerSet<T> {

    private static final int MAX_SIZE = 20;

    public InnerArraySet() {
        super(MAX_SIZE); // Optimise the array length to the size of InnerSet
    }

    public InnerArraySet(InnerSet<T> i) {
        this();
        for (InnerSetIterator<T> iterator = new InnerSetIterator<>(i); iterator.hasNext(); ) {
            add(iterator.next());
        }
    }

    public InnerArraySet(InnerSet<T> i, T p_element) {
        this(i);
        add(p_element);
    }

    @Override
    public InnerSet<T> addElement(T p_element) {
        if (!contains(p_element)) {
            if (getSize() >= MAX_SIZE) {
                return new InnerHashSet<T>(this, p_element);
            } else {
                add(p_element);
            }
        }
        return this;
    }

    @Override
    public InnerSet<T> addUnChecked(T p_element) {
        if (getSize() >= MAX_SIZE) {
            return new InnerHashSet<T>(this, p_element);
        } else {
            add(p_element);
            return this;
        }
    }

    @Override
    public InnerSet<T> addAllElements(InnerSet<T> innerSet) {
        // Reverse the add if the innerSet we add is smaller than this
        if (innerSet.getSize() < this.getSize()) {
            return innerSet.addAllElements(this);
        }
        for (T el : this) {
            innerSet = innerSet.addElement(el);
        }
        return innerSet;
    }

    @Override
    public boolean containsElement(Object p_element) {
        return contains(p_element);
    }

    @Override
    public boolean containsAllElements(InnerSet<T> innerSet) {
        return false;
    }

    @Override
    public T getElement(int index) {
        return get(index);
    }

    @Override
    public InnerSet<T> removeElement(Object p_element) {
        for (int index = 0; index < size(); index++) {
            if (p_element.equals(get(index))) {
                set(index, get(size() - 1));
                remove(size() - 1);
            }
        }
        return this;
    }

    @Override
    public InnerSet<T> removeAllElements(InnerSet<T> innerSet) {
        Iterator<T> it = innerSet.iterator();
        while (it.hasNext()) {
            removeElement(it.next());
        }
        return this;
    }

    @Override
    public InnerSet<T> retainAllElements(InnerSet<T> innerSet) {
        return null;
    }

    @Override
    public InnerSet<T> addAndPropagate(T p_element, MicroSet<T> microSetToPropagate) {
        if (!contains(p_element)) {
            microSetToPropagate.add(p_element);
            if (getSize() == MAX_SIZE) {
                return new InnerHashSet<>(this, p_element);
            } else {
                add(p_element);
            }
        }
        return this;
    }

    public InnerSet<T> addAllAndPropagate(InnerSet<T> innerSet, MicroSet<T> microSetToPropagate) {
        return innerSet.addAllAndPropagateReverse(this, microSetToPropagate);
    }

    @Override
    public InnerSet<T> addAllAndPropagateReverse(InnerSet<T> innerSet, MicroSet<T> microSetPropagate) {
        for (T el : this) {
            innerSet = innerSet.addAndPropagate(el, microSetPropagate);
        }
        return innerSet;
    }


    @Override
    public int getSize() {
        return size();
    }

    @Override
    public InnerSet<T> copy() {
        return new InnerArraySet<>(this);
    }

    @Override
    public InnerSet<T> clear(boolean unused) {
        this.clear();
        return this;
    }
}

File: src/scala/implementation/InnerArraySet.scala
package scala.implementation

import java.util.ArrayList

object InnerArraySet
{
  private val maxSize = 20
}

class InnerArraySet[T] extends ArrayList[T] with InnerSet[T] {

  def this(p_innerSet: InnerSet[T]) =
  {
    this
    val it: InnerSetIterator[T] = p_innerSet.iterator
    while(it.hasNext)
    {
      add(it.next())
    }
  }

  override def addElement(p_element: T): InnerSet[T] =
  {
    if(!contains(p_element))
    {
      if(getSize == InnerArraySet.maxSize){
        return new InnerHashSet[T](this, p_element)
      }
      else add(p_element)
    }
    this
  }

  override def addUnchecked(p_element: T): InnerSet[T] =
  {
    if(getSize == InnerArraySet.maxSize){
      return new InnerHashSet[T](this, p_element)
    }
    else add(p_element)
    this
  }

  override def addAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =
  {
    val it: InnerSetIterator[T] = this.iterator
    var innerSetReturn: InnerSet[T] = p_innerSet
      while(it.hasNext){
      innerSetReturn = innerSetReturn.addElement(it.next())
    }
    innerSetReturn
  }

  override def containsElement(p_element: Object): Boolean = contains(p_element)

  override def containsAllElements(p_innerSet: InnerSet[T]): Boolean =
  {
    val it: InnerSetIterator[T] = p_innerSet.iterator
    while(it.hasNext){
      if(!contains(it.next())) return false
    }
    true
  }

  override def getElement(index: Int): Option[T] = Option(get(index))

  override def removeElement(p_element: Object): InnerSet[T] =
  {
    var index = 0
    while (index < size) {
        if (p_element == get(index)) {
          set(index, get(size - 1))
          remove(size - 1)
        }
        index += 1
    }
    this
  }

  override def removeAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =
  {
    val it: InnerSetIterator[T] = p_innerSet.iterator
    while(it.hasNext){
      removeElement(it.next().asInstanceOf[Object])
    }
    this
  }

  override def retainAllElements(p_innerSet: InnerSet[T]): InnerSet[T] = ??? //TODO

  override def getSize: Int = size()

  override def iterator: InnerSetIterator[T] = new InnerSetIterator(this)

  override def copy: InnerSet[T] = new InnerArraySet(this)

  override def clear(unused: Boolean): InnerSet[T] = clear(unused)

  override def toString(): String =  "InnerArraySet : "+ this.toString()

  override def addAndPropagate(p_element: T, p_microSet: MicroSet[T]): InnerSet[T] =
  {
    if(!contains(p_element))
    {
      p_microSet.add(p_element)
      if(getSize == InnerArraySet.maxSize){
        return new InnerHashSet[T](this, p_element)
      }
      else add(p_element)
    }
    this
  }

  override def addAllAndPropagate(p_innerSet: InnerSet[T], p_microSet: MicroSet[T]): InnerSet[T] = p_innerSet.addAllAndPropagateReverse(this, p_microSet)

  override def addAllAndPropagateReverse(p_innerSet: InnerSet[T], p_microSet: MicroSet[T]): InnerSet[T] =
  {
    val it: InnerSetIterator[T] = this.iterator
    var innerSetReturn: InnerSet[T] = p_innerSet
      while(it.hasNext){
      innerSetReturn = innerSetReturn.addAndPropagate(it.next(),p_microSet)
    }
    innerSetReturn
  }
}
